### PR TITLE
42939 custom matchers

### PR DESCRIPTION
**Pull request by rhyshort**
Add custom matchers to match arrays if it contains all the specified elements. These are implemented as two matchers, one for whole queries and one for just arrays. 

The query matcher only checks top-level dict values for arrays, if the value is an array, it performs a check
to see if allElements are in the array, other wise it calls isEqual instead.

### CODE REVIEW

**Comment by mikerhodes** [comment]
Can you rebase this onto master so we can see if the build works?

**Comment by rhyshort** [comment]
@mikerhodes done

File: Example/Tests/Matchers/CDTQEitherMatcher.h [hunk 1]
```diff
@@ -0,0 +1,14 @@
+//
+//  CDTQEitherMatcher.h
+//  CloudantQueryObjc
+//
+//  Created by Rhys Short on 21/01/2015.
+//  Copyright (c) 2015 Michael Rhodes. All rights reserved.
```
Comment (mikerhodes) [inline Example/Tests/Matchers/CDTQEitherMatcher.h]: Copyright Cloudant :)

File: Example/Tests/Matchers/CDTQQueryMatcher.h [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+//
+//  Created by Rhys Short on 21/01/2015.
+//  Copyright (c) 2015 Michael Rhodes. All rights reserved.
+//
+
+#import <Foundation/Foundation.h>
+
+#import "Expecta.h"
+
+EXPMatcherInterface(beTheSameAs, (NSDictionary * expected));
```
Comment (mikerhodes) [inline Example/Tests/Matchers/CDTQQueryMatcher.h]: Add doc comments for each of the matchers to say what they do.

File: Pod/Classes/CDTQQueryExecutor.m [hunk 3]
```diff
@@ -197,7 +197,7 @@ + (BOOL)validateFields:(NSArray *)fields
             return NO;
         }
 
-        if ([field containsString:@"."]) {
+        if ([field rangeOfString:@"."].location != NSNotFound) {
```
Comment (mikerhodes) [inline Pod/Classes/CDTQQueryExecutor.m]: Good fix.

File: Example/Tests/Matchers/CDTQContainsAllElementsMatcher.h [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+//
+
+#ifndef CloudantQueryObjc_CDTQContainsAllElementsMatcher_h
+#define CloudantQueryObjc_CDTQContainsAllElementsMatcher_h
+
+#import "Expecta.h"
+
+EXPMatcherInterface(containsAllElements, (NSArray * expected));
+
+#define containAllElements containsAllElements
```
Comment (mikerhodes) [inline Example/Tests/Matchers/CDTQContainsAllElementsMatcher.h]: Is this the standard expecta way to do this? I assume this is here to make the case where one says `expect(blah, to(containAllElements(blah)` to make sense?

File: Example/Tests/Matchers/CDTQQueryMatcher.h [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+
+#define beTheSameQueryAs beTheSameAs
```
Comment (mikerhodes) [inline Example/Tests/Matchers/CDTQQueryMatcher.h]: I don't think we should have both of these defined. I suggest just keeping `beTheSameQueryAs` to keep the matcher's job more tightly specified.

**Comment by mikerhodes** [comment]
Looks generally good bar my minor quibbles.

File: Example/Tests/Matchers/CDTQContainsAllElementsMatcher.h (hunk 4 above)
Comment (rhyshort) [inline Example/Tests/Matchers/CDTQContainsAllElementsMatcher.h]: I think so, I was following the pattern in the expecta [readme] (https://github.com/specta/expecta/blob/master/README.md#writing-new-matchers), although the names could probably be improved

**Comment by mikerhodes** [comment]
+1 once squashed. You need a separate commit for:
- the changes to IBM Corp. in the licence headers.
- cf2e3d69cc11456fa8f95879098735844d13a688
- the main work on the new matchers.

**Pull request merged**